File: bnovate/bnovate/report/revenue_analytics_master/revenue_analytics_master.js
// Copyright (c) 2016, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Revenue Analytics Master"] = {
	filters: [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": new Date(new Date().getFullYear(), 0, 1), // First day of calendar year.
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": new Date(new Date().getFullYear(), 11, 31), // Last day of calendar year.
			"reqd": 1
		},
		{
			"fieldname": "include",
			"label": __("Include"),
			"fieldtype": "Select",
			"options": ["All", "Billed", "Unbilled"],
			"default": "All"
		},
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			fieldname: "revenue_streams",
			label: __("Revenue Streams"),
			fieldtype: "MultiSelectList",
			get_data: async function () {
				const revenue_streams = await frappe.db.get_list("Revenue Stream", {
					fields: ["name", "revenue_stream_name"],
					order_by: "lft",
					filters: {
						is_group: 0,
					}
				});
				return revenue_streams.map(rs => ({ value: rs.name, label: rs.revenue_stream_name }));
			},
			on_change: function () {
				frappe.query_report.refresh();
			},
		},

	],
	async onload(report) {
		// Select all revenue streams by default
		const stream_filter = frappe.query_report.get_filter("revenue_streams");
		if (stream_filter && !stream_filter.get_value()?.length) {
			const revenue_streams = await frappe.db.get_list("Revenue Stream", {
				fields: ["name"],
				order_by: "lft",
				filters: {
					is_group: 0,
				}
			});

			stream_filter.set_value(revenue_streams.map(rs => rs.name));
		}
	},
};

File: bnovate/bnovate/report/revenue_analytics_master/revenue_analytics_master.py
# Copyright (c) 2013, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

from frappe import _
from frappe.utils import flt, getdate, add_months
from erpnext import get_company_currency

 
def execute(filters=None):
    if not filters:
        filters = {}

    validate_filters(filters)

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data

def validate_filters(filters):
    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))

    if from_date > to_date:
        frappe.throw(_("From Date must be before To Date"))


def get_columns(filters):
    return [{
        "fieldname": "stage",
        "label": _("Stage"),
        "fieldtype": "Data",
        "width": 150
    }, 
    {
        "fieldname": "revenue_stream_name",
        "label": _("Revenue Stream"),
        "fieldtype": "Data",
        "width": 200
    },
    {
        "fieldname": "item_group",
        "label": _("Item Group"),
        "fieldtype": "Data",
        "width": 200
    },
    {
        "fieldname": "item_code",
        "label": _("Item Code"),
        "fieldtype": "Link",
        "width": 200,
        "options": "Item"
    },
    {
        "fieldname": "posting_date",
        "label": _("Posting Date"),
        "fieldtype": "Date",
        "width": 150
    },
    {
        "fieldname": "year",
        "label": _("Year"),
        "fieldtype": "Data",
        "width": 150
    },
    {
        "fieldname": "quarter",
        "label": _("Quarter"),
        "fieldtype": "Data",
        "width": 150
    },
    {
        "fieldname": "month",
        "label": _("Month"),
        "fieldtype": "Data",
        "width": 150
    },
    {
        "fieldname": "amount",
        "label": _("Amount"),
        "fieldtype": "Currency",
        "width": 150
    },
    {
        "fieldname": "so_name",
        "label": _("Sales Order"),
        "fieldtype": "Link",
        "options": "Sales Order",
        "width": 200
    },
    {
        "fieldname": "sinv_name",
        "label": _("Sales Invoice"),
        "fieldtype": "Link",
        "options": "Sales Invoice",
        "width": 200
    }]

def get_data(filters):
    filters.include_sinv = filters.get("include") in ("All", "Billed")
    filters.include_so = filters.get("include") in ("All", "Unbilled")

    filters.where_conditions = "WHERE true"    
    if filters.revenue_stream:
        filters.where_conditions += " AND IFNULL(rs.name, 'Other') = '{revenue_stream}' ".format(**filters)


    filters.shipping_default_account = frappe.get_value("Company", filters.company, "default_freight_sales_account")
    filters.shipping_item_group = "Shipping"

    data_query = """
        WITH orders AS (
            SELECT  -- SINV items
                sii.item_code,
                i.item_group as item_group,
                si.posting_date as posting_date,
                sii.base_net_amount as amount,
                sii.sales_order as so_name,
                si.name as sinv_name,
                "Billed" as stage
            FROM `tabSales Invoice` si
            JOIN `tabSales Invoice Item` sii ON sii.parent = si.name 
            JOIN `tabItem` i on i.item_code = sii.item_code
            
            WHERE si.docstatus = 1
                AND si.company = '{company}'
                AND si.posting_date BETWEEN '{from_date}' AND '{to_date}'
                AND {include_sinv}

            UNION ALL
            
            SELECT -- SINV Taxes and Charges, including Shipping
                NULL as item_code,
                IF(t.account_head = '{shipping_default_account}', '{shipping_item_group}', 'Taxes and Charges') as item_group,
                si.posting_date as posting_date,
                t.base_tax_amount as amount,
                NULL as so_name,
                si.name as sinv_name,
                "Billed" as stage
            FROM `tabSales Invoice` si
            JOIN `tabSales Taxes and Charges` t ON t.parent = si.name
            WHERE si.docstatus = 1
                AND si.company = '{company}'
                AND si.posting_date BETWEEN '{from_date}' AND '{to_date}'
                AND {include_sinv}

            UNION ALL

            SELECT  -- Unbilled SO items
                soi.item_code,
                i.item_group as item_group,
                soi.delivery_date as posting_date,
                (soi.net_amount - ifnull(soi.billed_amt, 0)) * so.conversion_rate AS amount,  -- Unbilled amount in company currency
                so.name as so_name,
                NULL as sinv_name,
                "Unbilled" as stage
            FROM `tabSales Order` so
            JOIN `tabSales Order Item` soi ON soi.parent = so.name 
            JOIN `tabItem` i on i.item_code = soi.item_code
            WHERE so.docstatus = 1
                AND so.status != 'Closed'
                AND so.company = '{company}'
                AND so.delivery_date BETWEEN '{from_date}' AND '{to_date}'
                AND(soi.net_amount - ifnull(soi.billed_amt, 0)) > 0
                AND {include_so}
        )

        SELECT
            o.stage,
            IFNULL(rs.name, 'Other') as revenue_stream_name,
            o.item_group as item_group,
            o.item_code,
            i.item_name,
            o.posting_date,
            YEAR(o.posting_date) as year,
            CONCAT(YEAR(o.posting_date), '-Q', QUARTER(o.posting_date)) as quarter,
            DATE_FORMAT(o.posting_date, '%Y-%m') as month,
            o.amount,
            o.so_name,
            o.sinv_name
        FROM orders o
        LEFT JOIN `tabItem` i on i.item_code = o.item_code
        LEFT JOIN `tabItem Group` ig ON ig.name = o.item_group
        LEFT JOIN `tabRevenue Stream` rs ON rs.name = ig.revenue_stream
        {where_conditions}
    """.format(**filters)
    
    aggregated_data = frappe.db.sql(data_query, as_dict=True)

    return aggregated_data